### PR TITLE
Safely read content from InputStream

### PR DESCRIPTION
**Pull request by resec**
Changed to use `ByteArrayOutputStream` on converting `InputStream` into `byte[]`, instead of relying on [`InputStream.available()`](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#available--).

The old way may cause problem in cases like reading from buffered input stream or a network stream.

### CODE REVIEW

**Review by asimshankar — CHANGES_REQUESTED** [review]
Thanks for the PR. Have some questions about peak memory usage.

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 1]
```diff
@@ -447,15 +448,13 @@ private void loadGraph(InputStream is, Graph g) throws IOException {
...
-              + numBytesRead
-              + " of the graph, expected to read "
-              + graphDef.length);
+    ByteArrayOutputStream baos = new ByteArrayOutputStream(16384);
+    int numBytesRead;
+    byte[] buf = new byte[16384];
+    while ((numBytesRead = is.read(buf, 0, buf.length)) != -1) {
+      baos.write(buf, 0, numBytesRead);
     }
+    byte[] graphDef = baos.toByteArray();
```
[the remark this answers is not in the document]
Comment (resec) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Thanks, it make sense. Let me rearrange the method calling.
Change the `loadGraph` to take `byte[]` and then in each constructor, build up the byte array in their own way.

**Comment by resec** [comment]
@asimshankar Hi, I updated as said in comment, also add a check to decide the byte array output stream init size. Please help reviewing the change again, thanks.

**Comment by yifeif** [comment]
@resec could you resolve the conflicts? Thank you!

**Review by asimshankar — APPROVED** [review]
Thanks! The change looks good, though I guess conflicts need to be resolved.

**Comment by resec** [comment]
@asimshankar conflict resolved, thanks.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by resec** [comment]
fix io exception not caught, not sure if i can trigger jenkins, let me try.

@tensorflow-jenkins test this please

**Comment by resec** [comment]
:( can not trigger full testing, please help to trigger the build again, thanks

**Comment by yifeif** [comment]
Thanks @resec! @tensorflow-jenkins test this please.

File: tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java [hunk 2]
```diff
@@ -78,10 +79,27 @@ public TensorFlowInferenceInterface(AssetManager assetManager, String model) {
...
     try {
-      loadGraph(is, g);
+      if (VERSION.SDK_INT >= 18) {
+        Trace.beginSection("initializeTensorFlow");
+        Trace.beginSection("readGraphDef");
+      }
+    
+      // TODO(ashankar): Can we somehow mmap the contents instead of copying them?
+      byte[] graphDef = new byte[is.available()];
+
```
Comment (asimshankar) [inline tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: I think we forgot to actually read the bytes into the `graphDef` array  here.
i.e., an:

`is.read(graphDef)`

is required?

Comment (asimshankar) [reply tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java]: Sent a fix in #13073

**Comment by resec** [comment]
@asimshankar thanks for the fix :)

**Pull request merged**